cpp/merkletree/sparse_merkle_tree.h

```diff
@@ -0,0 +1,243 @@
+#ifndef CERT_TRANS_MERKLETREE_SPARSE_MERKLE_TREE_H
+#define CERT_TRANS_MERKLETREE_SPARSE_MERKLE_TREE_H
+
+#include <bitset>
+#include <glog/logging.h>
+#include <stddef.h>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "merkletree/merkle_tree_interface.h"
+#include "merkletree/tree_hasher.h"
+
+class SerialHasher;
+
+const int kDigestSizeBits = 256;
+
+
+// TODO(alcutter): compare using bitset and BIGNUM.
+typedef std::bitset<kDigestSizeBits> Path;
+
+
+// Creates a Path from the bits passed in.
+Path PathFromBytes(const std::string& bytes);
+
+
+// Calculates the set of "null" hashes:
+// ...H(H(H("")||H(""))||H("")||(H(""))||...)...
+//
+// Visible out here because it's useful for testing too.
+std::vector<std::string> CalculateNullHashes(const TreeHasher& hasher);
+
+// Implementation of a Sparse Merkle Tree.
+//
+// The design is inspired by the tree described in
+// http://www.links.org/files/RevocationTransparency.pdf), but with some
+// of tweaks, most notably:
+//   1) Leaf values are hashed before being incorporated into the tree.
+//   2) Similar to the way it works in the CT MerkleTree, hashes are domain
+//      separated by prefixing the preimage with \x00 for leaves, and \x01 for
+//      internal nodes.  (This prevents generation of maps with identical root
+//      hashes by manipulating leaf data between neighbouring nodes.)
```

Comment:
@AlCutter explained it to me in person... He tells me that in the paper, it's just the hash of the two concatenated bits of data, so if you have two leafs under the same node that contain "hello" and "Al", you could change them to contain "hell" and "oAl", and all the hashes would remain the same.

Not sure if it really needs to be clarified *here*, since this is explaining how some *other* thing works?